### ISSUE TITLE
Driver dead-locks itself

### ISSUE THREAD

**Issue by NikolayJuly** [mongodb/mongo-swift-driver#457]
### Swift version
```
Apple Swift version 5.2.2 (swiftlang-1103.0.32.6 clang-1103.0.32.51)
Target: x86_64-apple-darwin19.4.0
```
### Operating system

```
Darwin Nikolays-iMac.home 19.4.0 Darwin Kernel Version 19.4.0: Wed Mar  4 22:28:40 PST 2020; root:xnu-6153.101.6~15/RELEASE_X86_64 x86_64
```

### Driver version

```
    {
        "package": "mongo-swift-driver",
        "repositoryURL": "https://github.com/mongodb/mongo-swift-driver.git",
        "state": {
          "branch": null,
          "revision": "81cd04b6e08cf9a3a1ef0435b996f5bc68fb1f7c",
          "version": "1.0.0-rc0"
        }
      },

```

### What is the version(s) of `mongod` that you are running with the driver?
```
db version v4.2.5
git version: 2261279b51ea13df08ae708ff278f0679c59dc32
allocator: system
modules: none
build environment:
    distarch: x86_64
    target_arch: x86_64
```

### What is your server topology?

For now I just start db in docker with default settings. 

## What is the problem? 

Driver dead-lock itself. 
I will attach code and screenshot of xcode, but in short - all threads of driver locked and waiting for `bson_mutex_lock (&pool->mutex);` 

How I did reproduce initially:
I just updated to new RC and implemented new functionality, which do massive read from DB for some report. 

As result, app just stop responding, and promises were never full-filled.

After this I created empty project started from Vapor 4 and added only mongo driver.
Issue can be reproduced only if you create many cursors. 

According to description - each cursor takes its own event loop. That would be fine, but they stuck in it and never completes. I would expect overall degradation of work of service if I creates 1024 cursors, since some of them will just wait for other cursors to complete. But non are complete its job, they just hang up. Which means after this - server is dead, non stop timeouts :(

Here is project code. To reduce size I deleted spm produced folder. So please execute `swift package generate-xcodeproj`
[SampleCode.zip](https://github.com/mongodb/mongo-swift-driver/files/4515980/SampleCode.zip)

What I need to do to reproduce the issue:
1. Generate test data - After run project you can just call `http://localhost:8080/createTestData`  on running service. It will create data to search from.
2. Read from this DB to lock driver. Trick here - create many read operations at the same time from different threads. Simulate many API calls to service from clients at the same time. To execute read call `http://localhost:8080/lockMongoDb` service

**Expected result**: Slow down on response, but complete all tasks and eventually return data. 
**Actual result**: No result for me at all. Before that, I saw at least one cursor complete its work and thats it, silence after. Nothing else happen. And driver is locked, no read operations can be done.

I can't attach just swift file. From zip all code was done in `routes.swift`, 2 funcs - one to generate test results, other to read them and lock app.

Screenshot of xcode, where all thread of driver are locked:
<img width="2267" alt="Screenshot 2020-04-22 at 12 47 36" src="https://user-images.githubusercontent.com/699116/79980883-ad450380-849b-11ea-8f03-08ba7e905bac.png">

And it looks like show stopper for me, since I can't go live, knowing that more than 5 requests at the same time might kill my service. (5 - number of events loop I see in debug)

Also I tried use sessions, but I didn't get how I can use them at all :( They won't help if requests come from different sessions, but might help is same session need to do massive read/write from db.

**Comment by kmahar** [comment]
Hi @NikolayJuly, thanks for reaching out and providing a detailed and thorough bug report!

I actually encountered what I believe was the same issue myself recently, and fixed it in 62ccaff2c0b12b70e2d65c0324fd418c0ec75b10.
Updating your example project to pin to that commit, I successfully reach the fatal error at the end of `lockMongoDb`. Can you try that out in your actual project and see if it fixes the issue?

We will be tagging 1.0.0-rc1 soon containing this fix.

**Comment by NikolayJuly** [comment]
@kmahar, Thank you a lot for your response. Yes, issue actually fixed. From today I will update releases page non-stop, waiting for rc1. 

Since I got your attention, I would like ask one more question, which I can't answer using google. This is about MongoDb in general. We have _id, which has timestamp of object creatio. But I can't see any way to query on this date :( So I have to duplicated it as `createdAt` field, if I want to query objects which created in specific time range.  Or did I miss smth?

Thanks for you help :) Feel free to close this issue, as it is resolved with your fix.

**Comment by kmahar** [comment]
I will aim to release it within the next week!

Closing the issue now.

Yes, there is a way to do what you are asking about. You might find [this blog post](https://kchodorow.com/2011/12/20/querying-for-timestamps-using-objectids/) helpful.

The first 4 bytes of an ObjectId contain a timestamp so you can construct ObjectIds corresponding to the start and end of your desired interval and query using those.

In Swift you can use `Date.timeIntervalSince1970` to convert `Date`s to seconds. You'll then need to convert those to hex strings and pad them with 0s for the remaining bytes. 

For example to construct one equivalent to Wednesday, April 22, 2020 7:39:05 PM GMT I would get the seconds (1587584345), convert them to hex (5ea09d59), and then use the `ObjectId` initializer accepting a hex string:
```swift
guard let start = ObjectId("5ea09d59" + "0000000000000000") else {
   // error, invalid formatting
}
```

And then you can filter your DB query using `$gt` and $lt`. 

That said, it is kind of a pain to do all that to construct the ObjectId. Some other drivers have methods to let you construct an ObjectId from a Date for exactly this purpose. I've opened a JIRA ticket [SWIFT-813](https://jira.mongodb.org/browse/SWIFT-813) suggesting we add one too.

**Comment by kmahar** [comment]
@NikolayJuly - after talking with my teammates, I'd actually advise against doing that little `ObjectId` trick. While it works currently, we don't guarantee that `ObjectId` generation will always be done in this way, and the byte generation methodology is meant to be opaque to the user.

Therefore I'd suggest one of the following:
- Continue using the `createdAt` timestamp as you mentioned 
- If you are using this collection for e.g. time series data and have a unique value in the collection for each point in time, you can use a different type (such as `Timestamp`) for your `_id` fields.

**Comment by NikolayJuly** [comment]
@kmahar  thanks for your information. I will not use your first proposal and will keep using separate `createdAt`. Thanks for your help!

**Comment by kmahar** [comment]
@NikolayJuly, we released 1.0.0-rc1 containing the deadlock fix earlier this week.

**Comment by NikolayJuly** [comment]
@kmahar thank you, for keeping me posted. Will update my services to use it.
